Added appearance settings

Diegogtz03 opened the pull request:
- Added ColorStore to store accent color preference in UserDefaults
- Populated the app with the preferred accent color in Notes & My Data
- Added some spacing in My Data to improve readability between the title and text
- Added multiple app icons
- Added setting to allow icon change

ericlewis commented on Ai Pin/Intents/EditNoteIntent.swift:
```diff
@@ -0,0 +1,31 @@
+struct EditNoteIntent: AppIntent {
+    static var title: LocalizedStringResource = "Edit Note"
+    
+    @Parameter(title: "Id") var id: String
```
```suggestion
    @Parameter(title: "Identifier") var id: String
```

ericlewis commented on Ai Pin/Stores/ColorStore.swift:
```diff
@@ -0,0 +1,27 @@
+import SwiftUI
+
+class ColorStore: ObservableObject {
```
Do not use ObservableObject, we use the new @Observable paradigm. This may require you factor things in multiple places.

Diegogtz03 replied:
Noted! Changed it to the new @observable paradigm just like the NavigationStore

ericlewis commented on Ai Pin/Views/MyDataView.swift:
```diff
@@ -34,11 +40,14 @@ struct MyDataView: View {
                             Text(p)
                                 .font(.headline)
+                                .foregroundStyle(colorStore.accentColor)
+                                .padding([.bottom], 5)
```
```suggestion
                                .padding(.bottom, 5)
```

ericlewis commented on Ai Pin/Views/Notes/NotesEditView.swift:
```diff
@@ -0,0 +1,63 @@
+struct NotesEditView: View {
+    @Environment(NavigationStore.self) private var navigationStore
+    @Environment(\.dismiss) private var dismiss
+    @State private var triggerSaveHatic = false
+    @State private var triggerCancelHatic = false
```
these don't appear to be doing anything.

Diegogtz03 replied:
They were in fact not doing anything 🤦🏻‍♂️. I removed the cancelled haptic and just kept the save haptic to provide some feedback.

ericlewis commented on Ai Pin/Views/Notes/NotesView.swift:
```diff
@@ -19,8 +21,9 @@ struct MemoryView: View {
                             }
                         }
-                    Text(note.text)
+                    Text(.init(note.text))
```
Is this so it will trigger markdown rendering?

ericlewis replied:
if so, it is preferred to use the proper named initializer.

Diegogtz03 replied:
Yes! It was intended to trigger markdown. However, I'm noticing that leaving Text() by itself also triggers markdown

ericlewis commented on Ai Pin/Views/Notes/NotesView.swift:
```diff
@@ -30,20 +33,23 @@ struct MemoryView: View {
     
-    @Environment(NavigationStore.self)
-    private var navigationStore
+    @State private var selectedNoteId = ""
+    @State private var selectedNote = Note(text: "", title: "")
+    
+    @Environment(NavigationStore.self) private var navigationStore
```
this will apply in a few places, but all property wrappers exist on their own line with what they are applied to below. This chunk of code should look like:

```swift
 @State 
 private var selectedNoteId = ""
 
 @State 
 private var selectedNote = Note(text: "", title: "")

 @Environment(NavigationStore.self) 
 private var navigationStore

Diegogtz03 replied:
Noted!

ericlewis commented on Ai Pin/Views/Notes/NotesView.swift:
```diff
@@ -68,6 +74,12 @@ struct NotesView: View {
                             .tint(.pink)
                         }
+                        .environmentObject(colorStore)
+                        .onTapGesture {
```
This should be a button not a tap gesture.

ericlewis commented on Ai Pin/Views/Settings/AccentColorView.swift:
```diff
@@ -0,0 +1,90 @@
+    @Environment(\.dismiss) private var dismiss
+    
+    var body: some View {
+        VStack(alignment: .leading) {
+            Text("This is an example title")
+                .foregroundStyle(accentColor)
+                .padding([.bottom], 5)
+                .bold()
+                .font(Font.system(size: 20))
+            Text("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum sem lacus, scelerisque ut ultrices in, aliquam ut massa. Vestibulum sed iaculis diam. In pulvinar felis nibh, ultrices bibendum tellus elementum ut. Ut non turpis feugiat, molestie justo sed.")
+                .foregroundStyle(.foreground)
+        }
+        .padding()
+        .background(.thinMaterial)
+        .clipShape(RoundedRectangle(cornerRadius: 25.0))
+        .padding(13)
+        .padding([.top], 20)
+        .onAppear() {
+            let tempColor = colorStore.loadColor()
+            accentColor = tempColor
+            originalColor = tempColor
+        }
```
if this is a "preview" it is better to use the actual view being previewed with fake data instead.

ericlewis commented on Ai Pin/Views/SettingsView.swift:
```diff
@@ -58,6 +62,23 @@ struct SettingsView: View {
+                                .environmentObject(colorStore)
+                        }
+                        
+                        Button("App Icon") {
+                            self.navigationStore.iconChangerPresented = true
+                        }
+                        .sheet(isPresented: $navigationStore.iconChangerPresented) {
+                            IconChangerView()
+                        }
```
you should only show this button for platforms that have icon changing.

ericlewis commented on Ai Pin/Views/Notes/NotesEditView.swift:
```diff
@@ -0,0 +1,63 @@
+    @State private var triggerSaveHatic = false
+    @State private var triggerCancelHatic = false
+    @State var noteId: String
+    @State var note: Note
```
you cannot instantiate @State from external sources. This is incorrect and leads to undefined behavior. What would be better would be for the note to be an observable of some sort, which I intend to do in SwiftDtcata. 

That said, to fix this for now you should use a custom initializer to store the note statically then set state in onAppear.

ericlewis commented on Ai Pin/API/API.swift:
```diff
@@ -227,7 +227,7 @@ extension API {
     }
     
-    func update(id: String, with note: Note) async throws -> Bool {
+    func update(id: String, with note: Note) async throws -> Any {
```
No, this is not correct. While the existing return type is wrong, it doesn't help to make it more wrong. The correct return type should be inspected and lined up with the appropriate model. In this case it is probably the`Memory` type.

ericlewis commented on Ai Pin/Stores/ColorStore.swift:
```diff
@@ -0,0 +1,27 @@
+    @Published var accentColor: Color = .blue;
+    
+    init() {
+        if let colorHex = UserDefaults.standard.object(forKey: "accentColor") {
```
don't use "magic strings" like this. There should be a Constants file somewhere which contains the key. Further, it is advised to prefix the key so it isn't so ambiguous and also to version it. "UI_CUSTOM_ACCENT_COLOR_V1".

Diegogtz03 replied:
Noted! Added an app constants file to manage this!

ericlewis reviewed (changes requested):
This is pretty good, thank you! There is a number of things that need fixing as commented in the PR itself and for some of the broader changes needed:

- For app icons the icon must be square and not have rounded corners at all. Please fill in the corners if possible. I am also not a *huge* fan of using humane's logo because it is a copyright of theirs, so please remove that. The app and repo will also be eventually renamed. 

- Style wise we always put property wrappers on their own line. 

- smaller more focused PRs (will help you land changes easier, since you won't have to rework parts that are fine)

- data management: do not use the "old" SwiftUI data flow, and you cannot under any circumstances ever set @State objects from `init`.

ericlewis commented:
I also didn't notice if you are setting an accentColor and tint for the whole app, but I noticed it wasn't on the edit note sheet, so I presume not. But it would be cool if the tint also affected the system components. You need only set tint / accentColor at the roots of sheet and app.

Diegogtz03 commented:
Thank you so much Eric! And definitely some work I have to do, really, thank you so much for the review! 

I'll change my up this code and consider all of this for any further changes in the future!

- About the Icons, I agree, I'll make sure to remove those affiliated directly with Humane and maybe just leave the "Sensors" one for now and leave the ability to add more icons in the future.
- And regarding the accentColor, yes! I currently pass it along to the specific titles, but retracted from adding it to the edit notes as to not mess much with the design, but will definitely make some cool changes and show you to see what you think! :)

Diegogtz03 commented:
Hi Eric! Just pushed the new changes, ty!!:

- I removed my changes regarding the notes updates
- Updated UI to match design for both appearance and icon changer
- Updated the preview view on accent color changer
- Populated selected accent color on the whole app
- Removed unwanted icons

ericlewis commented on Ai Pin/Ai_PinApp.swift:
```diff
@@ -4,20 +4,29 @@ import AppIntents
+        sceneColorStore = colorStore
 
         AppDependencyManager.shared.add(dependency: navigationStore)
+        AppDependencyManager.shared.add(dependency: colorStore)
```
```suggestion
```

ericlewis commented on Ai Pin/Stores/ColorStore.swift:
```diff
@@ -0,0 +1,27 @@
+import SwiftUI
+
+@Observable final class ColorStore: Sendable {
+    var accentColor: Color = .blue;
```
```suggestion
    var accentColor: Color;
```
